### PR TITLE
Update entitlement gated to finalize all roleids before clearing transaction

### PR DESCRIPTION
**Pull request by giuseppecrj**
This update enhances the entitlement check result handling in the EntitlementGatedBase contract:

**Result Posting and Event Emission:**
* Now occurs when the final status is PASSED OR when all role IDs have completed their checks.
* This ensures that results are posted and events are emitted for both successful checks and when all checks are complete, regardless of the outcome.

**Transaction Removal:**
* The transaction is now removed only when all role IDs have completed their checks.
* This change ensures that all necessary data is retained until the entire process is complete, improving data consistency and allowing for potential auditing or review of partial results.

These modifications provide a more comprehensive and flexible approach to handling entitlement check results, especially in scenarios involving multiple roles or complex permission structures.

### CODE REVIEW

**Review by clemire — CHANGES_REQUESTED** [review]

File: contracts/src/spaces/facets/gated/EntitlementGatedBase.sol
[the remark this answers is not in the document]
Comment (giuseppecrj) [reply contracts/src/spaces/facets/gated/EntitlementGatedBase.sol:1]: test cases added in the latest push.

File: contracts/src/spaces/facets/gated/EntitlementGatedBase.sol [hunk 1]
```diff
@@ -149,10 +149,34 @@ abstract contract EntitlementGatedBase is IEntitlementGatedBase {
       NodeVoteStatus finalStatus = passed > failed
         ? NodeVoteStatus.PASSED
         : NodeVoteStatus.FAILED;
-      _onEntitlementCheckResultPosted(transactionId, finalStatus);
-      emit EntitlementCheckResultPosted(transactionId, finalStatus);
-      _removeTransaction(transactionId);
+
+      bool allRoleIdsCompleted = _checkAllRoleIdsCompleted(transactionId);
+
+      if (finalStatus == NodeVoteStatus.PASSED || allRoleIdsCompleted) {
```
[the remark this answers is not in the document]
Comment (giuseppecrj) [reply contracts/src/spaces/facets/gated/EntitlementGatedBase.sol]: fixed in latest push.

**Comment by giuseppecrj** [comment]
Addresses on #1138

File: contracts/src/spaces/facets/gated/EntitlementGatedBase.sol [hunk 2]
```diff
@@ -146,13 +146,37 @@ abstract contract EntitlementGatedBase is IEntitlementGatedBase {
...
+      NodeVoteStatus finalStatusForRole = passed > failed
         ? NodeVoteStatus.PASSED
         : NodeVoteStatus.FAILED;
-      _onEntitlementCheckResultPosted(transactionId, finalStatus);
-      emit EntitlementCheckResultPosted(transactionId, finalStatus);
-      _removeTransaction(transactionId);
+
+      bool allRoleIdsCompleted = _checkAllRoleIdsCompleted(transactionId);
+
+      if (finalStatusForRole == NodeVoteStatus.PASSED || allRoleIdsCompleted) {
```
Comment (clemire) [inline contracts/src/spaces/facets/gated/EntitlementGatedBase.sol:155]: nice

File: contracts/test/crosschain/EntitlementGated.t.sol [hunk 3]
```diff
@@ -147,14 +167,86 @@ contract EntitlementGatedTest is
...
+    for (uint256 i; i < 3; ++i) {
+      vm.prank(nodes[i]);
+      gated.postEntitlementCheckResult(
+        requestId,
+        roleIds[0],
+        NodeVoteStatus.FAILED
+      );
+    }
+
+    // second roleId is entitled
```
Comment (clemire) [inline contracts/test/crosschain/EntitlementGated.t.sol:192]: nit - I think this comment is wrong, second role is also not entitled.

**Review by shuhuiluo — APPROVED** [review]

File: contracts/src/spaces/facets/gated/EntitlementGatedBase.sol [hunk 4]
```diff
@@ -112,11 +112,11 @@ abstract contract EntitlementGatedBase is IEntitlementGatedBase {
...
     uint256 failed = 0;
 
-    NodeVote[] storage nodeVotes = transaction.nodeVotesArray[roleId];
-    uint256 transactionNodesLength = nodeVotes.length;
+    NodeVote[] storage nodeVotesForRole = transaction.nodeVotesArray[roleId];
+    uint256 transactionNodesLength = nodeVotesForRole.length;
 
     for (uint256 i; i < transactionNodesLength; ++i) {
-      NodeVote storage tempVote = nodeVotes[i];
+      NodeVote storage tempVote = nodeVotesForRole[i];
```
[the remark this answers is not in the document]
Comment (giuseppecrj) [reply contracts/src/spaces/facets/gated/EntitlementGatedBase.sol:119]: will update in another PR

**Pull request merged**